unit_scaling/transforms/_simulate_format.py

```diff
@@ -0,0 +1,205 @@
+) -> None:
+    # Ideally we'd pass the formats as kwargs, but it currently causes a torch fx bug.
+    # This workaround will suffice for now...
+    args = [*node.args]
+    if len(node.args) == 2:  # pragma: no cover
+        args.append(None)
+    # Breaks when I pass in FPFormat objects, so convert to tuple and back
+    args = args[:3] + [fwd_format.to_tuple(), bwd_format.to_tuple()] + args[3:]
+
+    assert callable(node.target)
+    quantised_fn = _replacement_map[node.target]
+    logger.info("Quantising function: %s", node)
+    replace_node_with_function(graph, node, quantised_fn, args=tuple(args))
+
+
+def _quantisation_backend(fwd_format: FPFormat, bwd_format: FPFormat) -> Backend:
+    def backend_fn(gm: GraphModule, example_inputs: List[Tensor]) -> GraphModule:
+        logger.info("Running quantisation backend")
+        graph = gm.graph
+        for node in graph.nodes:
+            if node.op == "call_function" and node.target in _replacement_map:
+                _replace_with_quantised(graph, node, fwd_format, bwd_format)
+        return GraphModule(gm, graph)  # type: ignore[no-any-return]
+
+    return backend_fn
+
+
+def simulate_format(
+    module: nn.Module, fwd_format: FPFormat, bwd_format: FPFormat
+) -> nn.Module:
+    """[Experimental] Given a module, uses torch.dynamo to return a new module which
+    simulates the effect of using the supplied formats for matmuls.
+
+    Specifically, before each :func:`torch.nn.functional.linear` and
+    :func:`torch.nn.functional.scaled_dot_product_attention` call, a quantisation op
+    is inserted which simulates the effect of using the supplied `fwd_format`. This op
+    reduces the range of values to that of the given format, and (stochastically) rounds
+    values to only those representable by the format.
+
+    The same is true for the backward pass, where an op is inserted to quantise to the
+    `bwd_format`. Models which use modules that contain these functions internally
+    (such as :class:`torch.Linear`) will be inspected by torch.dynamo and have the
+    correct quantisation ops inserted.
+
+    If the equivalent unit-scaled functions from :mod:`unit_scaling.functional` are
+    used in the module, these too will be quantised.
+
+    Simulation of formats is run in FP32. Users should not expect speedups from using
+    this method. The purpose is to simulate the numerical effects of running matmuls
+    in various formats.
+
+    Args:
+        module (nn.Module): the module to be quantised
+        fwd_format (FPFormat): the quantisation format to be used in the forward pass
+            (activations and weights)
+        bwd_format (FPFormat): the quantisation format to be used in the backward pass
+            (gradients of activations and weights)
+
+    Returns:
+        nn.Module: a new module which when used, will run using the simulated formats.
+    """
+    module = deepcopy(module)
+    torch._dynamo.reset()  # type: ignore[no-untyped-call]
```

Comment:
What happens if you don't do that?

Reply:
Torch dynamo throws a warning if you call it with a different backend without calling that first, and may (I'm not sure) not actually use the new backend. Not sure what's going on there, but I don't think this hurts in our case.